### PR TITLE
chainntnfs: Reorder commit hints and notify

### PR DESCRIPTION
**Pull request by cfromknecht**
This PR modifies the chainntnfs package to properly commit height hints before delivering notifications. The primary motiviation is to ensure that downstream clients are provided with an externally consistent and crash consistent view of the notifier.

This PR comprises the first four commits in #1787

### CODE REVIEW

File: chainntnfs/btcdnotify/btcd.go [hunk 1]
```diff
@@ -772,21 +762,51 @@ func (b *BtcdNotifier) handleBlockConnected(epoch chainntnfs.BlockEpoch) error {
...
+	// so if any subscribers make queries based on their received
+	// block epoch, our state is fully updated in time.
+	b.bestBlock = epoch
+
+	// Next we'll notify any subscribed clients of the block.
+	b.notifyBlockEpochs(int32(newBlock.height), &newBlock.hash)
+
+	// Finally, send off the spend details to the notification subscribers.
+	for prevOut, batch := range spendBatches {
+		delete(b.spendNotifications, prevOut)
```
Comment (cfromknecht) [inline chainntnfs/btcdnotify/btcd.go]: this line can be removed

File: chainntnfs/neutrinonotify/neutrino.go [hunk 2]
```diff
@@ -645,42 +648,63 @@ func (n *NeutrinoNotifier) handleBlockConnected(newBlock *filteredBlock) error {
...
+	// block epoch, our state is fully updated in time.
+	n.bestHeight = newBlock.height
+
+	// With all persistent changes committed, notify any subscribed clients
+	// of the block.
+	n.notifyBlockEpochs(int32(newBlock.height), &newBlock.hash)
+
+	// Finally, send off the spend details to the notification subscribers.
+	for prevOut, batch := range spendBatches {
+		delete(n.spendNotifications, prevOut)
```
Comment (cfromknecht) [inline chainntnfs/neutrinonotify/neutrino.go]: this can be moved up to mirror the btcd change

File: chainntnfs/neutrinonotify/neutrino.go [hunk 3]
```diff
@@ -645,42 +648,63 @@ func (n *NeutrinoNotifier) handleBlockConnected(newBlock *filteredBlock) error {
...
 
-	// Finally, we'll update the spend height hint for all of our watched
+	// Now, we'll update the spend height hint for all of our watched
 	// outpoints that have not been spent yet. This is safe to do as we do
 	// not watch already spent outpoints for spend notifications.
 	ops := make([]wire.OutPoint, 0, len(n.spendNotifications))
 	for op := range n.spendNotifications {
-		ops = append(ops, op)
+		// Skip any outpoints that are to be triggered by this block.
+		if _, ok := spendBatches[op]; !ok {
```
Comment (cfromknecht) [inline chainntnfs/neutrinonotify/neutrino.go]: makes this unnecessary

File: chainntnfs/btcdnotify/btcd.go [hunk 4]
```diff
@@ -772,21 +762,51 @@ func (b *BtcdNotifier) handleBlockConnected(epoch chainntnfs.BlockEpoch) error {
 	// not watch already spent outpoints for spend notifications.
 	ops := make([]wire.OutPoint, 0, len(b.spendNotifications))
 	for op := range b.spendNotifications {
-		ops = append(ops, op)
+		// Skip any outpoints that are to be triggered by this block.
```
Comment (cfromknecht) [inline chainntnfs/btcdnotify/btcd.go]: reminder to remove this as well

**Comment by wpaulino** [comment]
FWIW most of this logic will no longer live within the different backend `ChainNotifier`s when merging the handling of spend notifications into the `TxConfNotifier`, so it might not be worth the trouble of getting this in.

**Comment by cfromknecht** [comment]
Cool with me, I can drop these commits from #1787

**Comment by cfromknecht** [comment]
Perhaps only the last commit is relevant then?

**Comment by cfromknecht** [comment]
@halseth, @wpaulino have decided it's best to keep this PR since there are upwards of 50 commits that build on them, and we'd like to minimize the chance of error in rebasing. The changes as they are fix existing issues in the code base. Getting them in at least ensures the issues do not arise while the upstream PRs are reviewd.

File: chainntnfs/btcdnotify/btcd.go [hunk 5]
```diff
@@ -566,7 +566,7 @@ func (b *BtcdNotifier) confDetailsFromTxIndex(txid *chainhash.Hash,
 	// then we may be able to dispatch it immediately.
 	tx, err := b.chainConn.GetRawTransactionVerbose(txid)
 	if err != nil {
-		// If the transaction lookup was successful, but it wasn't found
+		// If the transaction lookup was succesful, but it wasn't found
```
Comment (halseth) [inline chainntnfs/btcdnotify/btcd.go]: _googles "succesful_. Did you mean: "successful"?

File: chainntnfs/neutrinonotify/neutrino.go [hunk 6]
```diff
@@ -630,12 +633,13 @@ func (n *NeutrinoNotifier) handleBlockConnected(newBlock *filteredBlock) error {
 
 			// If this transaction indeed does spend an output which
 			// we have a registered notification for, then create a
-			// spend summary, finally sending off the details to the
-			// notification subscriber.
+			// spend summary and add it to our batch of spend
+			// notifications to be delivered.,
```
Comment (halseth) [inline chainntnfs/neutrinonotify/neutrino.go]: nit: end of sentence

File: chainntnfs/neutrinonotify/neutrino.go [hunk 7]
```diff
@@ -810,11 +831,11 @@ func (n *NeutrinoNotifier) RegisterSpendNtfn(outpoint *wire.OutPoint,
 	}
 
 	// Ensure that neutrino is caught up to the height hint before we
-	// attempt to fetch the utxo fromt the chain. If we're behind, then we
+	// attempt to fetch the utxo from the chain. If we're behind, then we
 	// may miss a notification dispatch.
 	for {
 		n.heightMtx.RLock()
-		currentHeight = n.bestHeight
+		currentHeight := n.bestHeight
```
Comment (halseth) [inline chainntnfs/neutrinonotify/neutrino.go]: wasn't this fixed already? Might need a rebase.

File: chainntnfs/btcdnotify/btcd.go (hunk 5 above)
Comment (cfromknecht) [reply chainntnfs/btcdnotify/btcd.go]: fixed

File: chainntnfs/neutrinonotify/neutrino.go (hunk 6 above)
Comment (cfromknecht) [reply chainntnfs/neutrinonotify/neutrino.go]: fixed

File: chainntnfs/neutrinonotify/neutrino.go (hunk 7 above)
Comment (cfromknecht) [reply chainntnfs/neutrinonotify/neutrino.go]: fixed

Comment (halseth) [reply chainntnfs/neutrinonotify/neutrino.go]: This was just change here: https://github.com/lightningnetwork/lnd/pull/1953

Why change it back?

**Pull request merged**